R/backup_planning.R

```diff
@@ -0,0 +1,73 @@
+# This file is supposed to be run to create timed backups of the MNM planning google sheet.
+# The table is dumped to a binary file.
+# Pure backup for now; we will take care of restoration later.
+
+
+source("R/functions.R")
+library(googlesheets4)
+
+#' read the Planning_v2 sheet in the planning googlesheet.
+#' @details see R/functions.R/get_planning_long()
+#' @param ss The id of the planning googlesheet
+get_planning_raw <- function(ss = gs_id()) {
+  # read the planning data
+  read_sheet(
+    ss,
+    sheet = "Planning_v2",
+    col_types = "ccccclllilccccccdddddddddddccdddddddddddcc",
+    .name_repair = "minimal"
+    )
+}
+
+
+#' Procedure to store a local backup of the google planning table.
+#' @details
+#' This will take a table and save it to disk,
+#' either as `rds` or as `csv`.
+#' @param ss a google sheet ID.
+#' @param filetype rds (binary) or csv (text).
+#' @param verbose whether or not output shall be printed.
+#' @examples local_backup(verbose = TRUE, filetype = 'rds')
+local_backup <- function(ss = gs_id(), verbose = TRUE, filetype = c('rds', 'csv')) {
+
+  # query the planning table
+  planning_table <- get_planning_raw(ss)
+
+  # check if the directory exists
+  if (!file.exists(here::here("local_backups"))) {
+    if (utils::askYesNo("Shall we create a local backups folder?")) {
+      # create it... if the user confirms
+      dir.create(here::here("local_backups"), showWarnings = FALSE)
+    } else {
+      print("No backup stored.")
+      return(NA)
+    }
+  }
+
+  # backup the table
+  if (missing(planning_table)) {
```

Comment:
AFAIK `missing()` is only used for function arguments.

`get_planning_raw(ss)` will error if it can't get the table. But there will be ways to capture that and provide a more informative error, if that was your aim.

Reply:
This is a tricky one... I think I had a situation when the google sheet would not load and the function returned `NA`. Yet I could not use `is.na()` because that works on data frames. 
`exists` evaluates T for `x <- NA`; and `is.null` is not appropriate either.
yet you are right: the base-r docs reserve `missing` for args.

Reply:
> situation when the google sheet would not load and the function returned `NA`

I can't reproduce that unfortunately.

But to check for a _scalar_ `NA`, you could do `is.atomic(planning_table) && length(planning_table) == 1 && is.na(planning_table)`. Perhaps you can use this to replace `missing()` (need to change something to it anyway).

The `is.atomic()` would be to exclude data frames from this test (data frames are lists of columns, so they have a length). One could also extend and check for empty dataframes:

```r
(is.atomic(planning_table) &&
  length(planning_table) == 1 &&
  is.na(planning_table)) ||
  (inherits(planning_table, "data.frame") && nrow(planning_table) == 0)
```

Reply:
Oh, wow, "atomic" seems to be a weird concept to me. 
I am hesitant to include this highly specific `(... && ... && ...)` check, and it has some overlap with the `data.frame` check.
I will try to rather positively evaluate whether we have a `data.frame` derivate; and in a second check, its dimensions.

Reply:
Yeah, you solved that very elegantly! Thanks.